### PR TITLE
Add row conv operator

### LINKED ISSUE #5612: Row Convolution Operator.
- [RowConvLayer.cpp]( https://github.com/PaddlePaddle/Paddle/blob/develop/paddle/gserver/layers/RowConvLayer.cpp)
- Introduced in DeepSpeech2 : http://www.cs.cmu.edu/~dyogatam/papers/wang+etal.iclrworkshop2016.pdf

### PR DESCRIPTION
**Pull request by sidgoyal78**
Resolves #5612 , by adding the implementation of the row-convolution operator.

Few notes:
- The CPU version implementation is very straight-forward as of now (it has unvectorized loops) and can be optimized.
- The GPU version is a direct porting of kernels written by @qingqing01 for the earlier version of paddle.(https://github.com/PaddlePaddle/Paddle/blob/develop/paddle/function/RowConvOpGpu.cu). I took some time to see if I could improve something, but could not think of improvements.

### CODE REVIEW

File: paddle/operators/row_conv_op.cc [hunk 1]
```diff
@@ -0,0 +1,246 @@
...
+      for (int k = 0; k < current_timesteps;
+           k++) {  // For different time steps in the same sequence
+        for (int w = 0; (w < context_length) && ((k + w) < current_timesteps);
+             w++) {
+          for (int d = 0; d < input_dim; d++) {
+            if (w == 0) {
+              cot_seq(k, d) = weights(w, d) * cip_seq(k + w, d);
+            } else {
+              cot_seq(k, d) += weights(w, d) * cip_seq(k + w, d);
+            }
```
Comment (qingqing01) [inline paddle/operators/row_conv_op.cc]: Maybe we can use elementwise mul and col-wise sum to remove the `for loop` in line 145 and line 147. But the optimization can be done in the future. So in this PR, I think it is ok here.

File: paddle/operators/row_conv_op.cc [hunk 2]
```diff
@@ -0,0 +1,246 @@
...
+  }
+};
+
+template <typename T>
+class RowConvKernel<platform::CPUPlace, T> : public framework::OpKernel<T> {
+ public:
+  void Compute(const framework::ExecutionContext &context) const override {
+    auto *X = context.Input<LoDTensor>("X");
+    auto *Filter = context.Input<Tensor>("Filter");
+    auto *Out = context.Output<LoDTensor>("Out");
```
Comment (qingqing01) [inline paddle/operators/row_conv_op.cc]: The naming style:
X->x
Filer->filter
Out->out

https://google.github.io/styleguide/cppguide.html#Variable_Names

File: paddle/operators/row_conv_op.cc [hunk 3]
```diff
@@ -0,0 +1,246 @@
...
+    AddInput("X",
+             "(LoDTensor), the input(X) is a LodTensor, which supports "
+             "variable time-length input sequences. The underlying tensor "
+             "in this LoDTensor is a matrix with shape (T x N), where T "
+             "is the total time steps in this mini-batch and N is the input "
+             "data dimension.");
+    AddInput("Filter",
+             "(Tensor), the input(Filter) is a learnable parameter. It "
+             "is a 2-D tensor with shape (future_context x N), where, "
+             "future_context is the batch size and N is the data dimension.");
```
[the remark this answers is not in the document]
Comment (sidgoyal78) [reply paddle/operators/row_conv_op.cc]: Fixed, thanks.

File: paddle/operators/row_conv_op.cc [hunk 4]
```diff
@@ -0,0 +1,246 @@
...
+
+This operator was introduced in the paper:
+http://www.cs.cmu.edu/~dyogatam/papers/wang+etal.iclrworkshop2016.pdf
+Given an input sequence $in$ of length $t$ and input dimension $d$, 
+and a filter ($W$) of size $context \times d$, 
+the output sequence is convolved in the following manner:
+
+$$
+out_{i, :} = \sum_{j=i}^{i + context} in_{j,:} \dot W_{i-j, :}
+$$
```
[the remark this answers is not in the document]
Comment (sidgoyal78) [reply paddle/operators/row_conv_op.cc]: Thanks for the pointer, i have included now.

File: paddle/operators/row_conv_op.cc (hunk 2 above)
Comment (sidgoyal78) [reply paddle/operators/row_conv_op.cc]: Fixed.

File: paddle/operators/row_conv_op.cc [hunk 5]
```diff
@@ -0,0 +1,246 @@
...
+template <typename T>
+class RowConvKernel<platform::CPUPlace, T> : public framework::OpKernel<T> {
+ public:
+  void Compute(const framework::ExecutionContext &context) const override {
+    auto *X = context.Input<LoDTensor>("X");
+    auto *Filter = context.Input<Tensor>("Filter");
+    auto *Out = context.Output<LoDTensor>("Out");
+
+    Out->mutable_data<T>(context.GetPlace());
+    context.ShareLoD("X", "Out");
```
[the remark this answers is not in the document]
Comment (sidgoyal78) [reply paddle/operators/row_conv_op.cc]: Fixed.

File: paddle/operators/row_conv_op.cc (hunk 2 above)
Comment (sidgoyal78) [reply paddle/operators/row_conv_op.cc]: (I need to fix for .cu code)

File: paddle/operators/row_conv_op.cc (hunk 1 above)
Comment (sidgoyal78) [reply paddle/operators/row_conv_op.cc]: Sure.

**Pull request merged**